Emit an 'active' metric for events to address issue #7667

gwmullin opened the pull request:
This address https://github.com/blakeblackshear/frigate/issues/7667 - The desire to be able to determine the count of active vs inactive objects (I want this for hass automation, to turn on my house number when a new car drives up on the street)

Rather than emit stationary objects, I implemented support for tracking active objects and left the existing signals alone to ensure backwards compatibility with any existing functionality people are using.

gwmullin committed:
Initial implementation of active object counters. Need to clean up a bit more and examine reuse of stationary/active logic in neighboring modules.

gwmullin committed:
A bit more cleanup for references to active, referencing the tracked object method rather than duplicating logic.

gwmullin committed:
Minor formatting and readability cleanup

gwmullin committed:
Update docs with the new active mqtt metric definition.

NickM-27 commented on frigate/object_processing.py:
```diff
@@ -165,6 +166,10 @@ def update(self, current_frame_time, obj_data):
         if self.computed_score > self.top_score:
             self.top_score = self.computed_score
         self.false_positive = self._is_false_positive()
+        if not self.active == self.is_active(): 
```
this is incorrect, it can only be a significant change if the object is not a false positive, this condition should stay where it was before

gwmullin replied:
Lost my comment - I stored this in a variable and moved the check back into the code block protected by the false_positive check.

gwmullin committed:
Move the check for a change in active status into the code block protected by a false positive check.

gwmullin reviewed (commented):
Fixed.

NickM-27 commented on frigate/object_processing.py:
```diff
@@ -165,6 +166,9 @@ def update(self, current_frame_time, obj_data):
             self.top_score = self.computed_score
         self.false_positive = self._is_false_positive()
+        # Hold previous active state for checking further down.
+        previously_active = self.active
```
I don't think it makes sense to use a separate variable for this, it can be stored as part of the previous object data, right?

hawkeye217 commented:
You'll need to format the code with `ruff` so the tests can pass.

gwmullin committed:
- Add 'active' to the tracked object dictionary, use the previous object for active comparison.

  - I also missed emitting updates when a tracked object is no longer tracked, and added handling for emitting zeros on object types.

gwmullin commented:
> You'll need to format the code with `ruff` so the tests can pass.

Done - thanks. I'll do a separate pull req to update the development docs and reference that.

NickM-27 commented:
there should be no reason to close this

gwmullin commented:
Oops wrong button, sorry! Re-opening req - I reworked the logic to use the previous object for comparison, and fixed the missing handling of 'end', so things are now updating as expected. I've tested locally with and without zones and things look to be in good shape now.

NickM-27 commented:
this will be pinned until 0.14.1 is released and 0.15 is merged into dev

gwmullin commented:
> this will be pinned until 0.14.1 is released and 0.15 is merged into dev

Great - thanks for the review, you all are doing amazing work, seemed only fair to tackle something I wanted from the project :)
